lib/internal/errors.js

```diff
@@ -737,6 +737,8 @@ E('ERR_NAPI_INVALID_TYPEDARRAY_ALIGNMENT',
   'start offset of %s should be a multiple of %s', RangeError);
 E('ERR_NAPI_INVALID_TYPEDARRAY_LENGTH',
   'Invalid typed array length', RangeError);
+E('ERR_NAPI_THREADSAFE_FUNCTION', 'N-API Threadsafe function operation failed',
+  Error);
```

Comment:
Generally the errors are more specific and there would be a different one for each of the errors. @jasnell what is your take in this case?

Reply:
Hmm... I'd prefer if the error message carried more detail about what the actual failure was

Reply:
I need some guidance then. This is the error code. In addition to that, there is also a message. If I make the code very specific, what is there left for the message?

For example, at [0](https://github.com/nodejs/node/blob/master/src/node_api.cc#L153-L156) we have
```C++
    THROW_RANGE_ERROR_IF_FALSE((env), (length) * (size_of_element) +           \
        (byte_offset) <= buffer->ByteLength(),                                 \
        "ERR_NAPI_INVALID_TYPEDARRAY_LENGTH",                                  \
"Invalid typed array length"); \
```
whereas at [1](https://github.com/nodejs/node/blob/master/lib/internal/errors.js#L738-L739) we have
```JS
E('ERR_NAPI_INVALID_TYPEDARRAY_LENGTH',
'Invalid typed array length', RangeError);
```

The message of the latter duplicates the message of the former. So, if every error code has a specific message associated with it in lib/internal/errors.js, then what is the point of the second parameter to `napi_throw_range_error(napi_env env, const char* code, const char* message)` which becomes the constructor parameter to the corresponding v8::Exception::Error?

Reply:
I think what we want are distinct error codes for each of the failure modes here, and a single message (or message template) associated with each one of them.

> This is the error code. In addition to that, there is also a message. If I make the code very specific, what is there left for the message?

The original motivation for introducing error codes was making the error message something only intended for humans, so that all programmatically available information would already be contained in the error code or other properties of the error object.

In particular, what we want to avoid is the need for API users to parse error messages to find out which of the failure conditions was hit exactly? That allows us to change them as we see fit, without it being a semver-major change, and maybe even translate them in the future.

Reply:
It looks like for pretty much all `NAPI_ERR_*` codes we duplicate the message from lib/internal/errors.js in src/node_api.cc ... so, I guess I'll do the same.

Reply:
@addaleax roger that! Will do. Thanks!